### PR TITLE
All types of the 'Function' constructor calls wrapped (close #769)

### LINKED ISSUE #769: Not all types of 'Function' constructor calls wrapped
wrapped: `new Function(...)`
non-wrapped: `Function(...)`, `Function.call(...)` and `Function.apply(...)`

see: T423261

### PR DESCRIPTION
**Pull request by LavrovArtem**
/cc @churkin @miherlosev

### CODE REVIEW

File: src/client/sandbox/code-instrumentation/index.js [hunk 1]
```diff
@@ -53,19 +53,33 @@ export default class CodeInstrumentation extends SandboxBase {
...
+                            args.push(script[i]);
+
+                        args.push(processScript(script[len - 1], false, false));
+
+                        return args;
+                    }
+                    else if (!isLast && typeof script[0] === 'string') {
+                        args.push(processScript(script[0], false, false));
+
+                        for (i = 1; i < len; i++)
```
Comment (inikulin) [inline src/client/sandbox/code-instrumentation/index.js]: Can we refactor this? I have no idea wht's going on here now?

**Comment by LavrovArtem** [comment]
FPR

**Comment by miherlosev** [comment]
I think you should use another way to support this scenario - override window.Function constructor.
For example see https://github.com/DevExpress/testcafe-hammerhead/blob/master/src/client/sandbox/node/window.js
\r-

**Comment by LavrovArtem** [comment]
@miherlosev Thanks, good idea

**Comment by LavrovArtem** [comment]
FPR

**Comment by LavrovArtem** [comment]
ping

File: src/client/sandbox/native-methods.js [hunk 2]
```diff
@@ -163,6 +163,7 @@ class NativeMethods {
         this.Blob             = win.Blob || mock;
         this.XMLHttpRequest   = win.XMLHttpRequest || mock;
         this.Image            = win.Image || mock;
+        this.Function         = win.Function || mock;
```
Comment (miherlosev) [inline src/client/sandbox/native-methods.js]: Why do we need `|| mock` instruction?

File: src/client/sandbox/node/window.js [hunk 3]
```diff
@@ -222,6 +222,16 @@ export default class WindowSandbox extends SandboxBase {
         };
         window.Image.prototype = nativeMethods.Image.prototype;
 
+        window.Function = function (...args) {
+            var lastArgIndex = args.length - 1;
+
```
Comment (miherlosev) [inline src/client/sandbox/node/window.js]: More readable name - `functionBodyArgIndex`

File: src/client/sandbox/node/window.js [hunk 4]
```diff
@@ -222,6 +222,16 @@ export default class WindowSandbox extends SandboxBase {
...
+            if (typeof args[lastArgIndex] === 'string')
+                args[lastArgIndex] = processScript(args[lastArgIndex], false, false);
+
+            return nativeMethods.Function.apply(this, args);
+        };
+        window.Function.prototype = nativeMethods.Function.prototype;
+
```
Comment (miherlosev) [inline src/client/sandbox/node/window.js]: Add test for this line. See https://github.com/DevExpress/testcafe-hammerhead/commit/589d08b0aed68e108168b2d4bae5df10fdea271c

File: test/client/fixtures/sandbox/node/classes-test.js [hunk 5]
```diff
@@ -210,3 +210,20 @@ if (navigator.registerProtocolHandler) {
...
 }
 
+test('window.Function must be overriden (T423261) (GH-769)', function () {
+    var removeOuterFunctionRe = /^[^{]+\{\s+|\s+}$/g;
+    var fnCtor                = Function;
+    var codeStr               = 'location.href="/page.htm";';
+    var processedCodeStr      = processScript(codeStr).trim();
+
+    /*eslint-disable no-new-func*/
+    strictEqual(new Function('arg1', 'arg2', codeStr).toString().replace(removeOuterFunctionRe, ''), processedCodeStr);
```
Comment (churkin) [inline test/client/fixtures/sandbox/node/classes-test.js]: `.toString().replace(removeOuterFunctionRe, '')` 
use a separate method

File: src/client/sandbox/node/window.js [hunk 6]
```diff
@@ -222,6 +222,16 @@ export default class WindowSandbox extends SandboxBase {
         };
         window.Image.prototype = nativeMethods.Image.prototype;
 
+        window.Function = function (...args) {
+            var lastArgIndex = args.length - 1;
+
+            if (typeof args[lastArgIndex] === 'string')
+                args[lastArgIndex] = processScript(args[lastArgIndex], false, false);
+
+            return nativeMethods.Function.apply(this, args);
```
Comment (miherlosev) [inline src/client/sandbox/node/window.js]: Add test that check throwing exceptions. 
See example - https://github.com/DevExpress/testcafe-hammerhead/blob/master/test/client/fixtures/sandbox/node/document-test.js#L260

**Comment by miherlosev** [comment]
\r-

File: src/client/sandbox/node/window.js (hunk 4 above)
Comment (LavrovArtem) [inline src/client/sandbox/node/window.js]: already done https://github.com/DevExpress/testcafe-hammerhead/pull/774/files#diff-e800e680b15ed48d6d4fa78f33822187R226

Comment (miherlosev) [inline src/client/sandbox/node/window.js]: Ok

File: src/client/sandbox/native-methods.js (hunk 2 above)
Comment (LavrovArtem) [inline src/client/sandbox/native-methods.js]: All window's classes use this instruction

Comment (miherlosev) [inline src/client/sandbox/native-methods.js]: @churkin 
Do you remember why we need this instruction?
I remove this and run qunit tests in IE11 and Chrome. All tests are green.

File: test/client/fixtures/sandbox/node/classes-test.js (hunk 5 above)
Comment (LavrovArtem) [inline test/client/fixtures/sandbox/node/classes-test.js]: ok

File: src/client/sandbox/node/window.js (hunk 3 above)
Comment (LavrovArtem) [inline src/client/sandbox/node/window.js]: ok

File: src/client/sandbox/native-methods.js (hunk 2 above)
Comment (churkin) [inline src/client/sandbox/native-methods.js]: @miherlosev Try to look at history

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
wait for #792

File: src/client/sandbox/node/window.js (hunk 6 above)
Comment (LavrovArtem) [inline src/client/sandbox/node/window.js]: Any of window's classes don't check. If you think that it needs, you can make separate issue to check all classes

**Comment by LavrovArtem** [comment]
FPR

File: src/client/sandbox/native-methods.js (hunk 2 above)
Comment (miherlosev) [inline src/client/sandbox/native-methods.js]: After discussing with @churkin we decide to remove `|| mock`.

File: src/client/sandbox/node/window.js (hunk 4 above)
Comment (miherlosev) [inline src/client/sandbox/node/window.js]: I create a separate ticke #793

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Comment by LavrovArtem** [comment]
FPR

File: src/client/sandbox/node/window.js [hunk 7]
```diff
@@ -222,6 +222,16 @@ export default class WindowSandbox extends SandboxBase {
         };
         window.Image.prototype = nativeMethods.Image.prototype;
 
+        window.Function = function (...args) {
+            var functionBodyArgIndex = args.length - 1;
+
+            if (typeof args[functionBodyArgIndex] === 'string')
+                args[functionBodyArgIndex] = processScript(args[functionBodyArgIndex], false, false);
```
Comment (miherlosev) [inline src/client/sandbox/node/window.js]: `processScript` has only 2 parameters

**Comment by LavrovArtem** [comment]
@testcafe-build-bot \retest

**Pull request merged**